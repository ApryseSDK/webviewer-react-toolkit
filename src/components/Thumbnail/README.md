A thumbnail representation of a PDF page.
