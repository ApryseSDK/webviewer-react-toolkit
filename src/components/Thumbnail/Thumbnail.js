'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
exports.Thumbnail = void 0;
var tslib_1 = require('tslib');
var classnames_1 = tslib_1.__importDefault(require('classnames'));
var react_1 = tslib_1.__importDefault(require('react'));
var hooks_1 = require('../../hooks');
var ClickableDiv_1 = require('../ClickableDiv');
var EditableText_1 = require('../EditableText');
var FilePlaceholder_1 = require('../FilePlaceholder');
var FileSkeleton_1 = require('../FileSkeleton');
var Image_1 = require('../Image');
var ToolButton_1 = require('../ToolButton');
var thumbnailFocusObservable_1 = require('./thumbnailFocusObservable');
function Thumbnail(_a) {
  var file = _a.file,
    label = _a.label,
    selected = _a.selected,
    dragging = _a.dragging,
    otherDragging = _a.otherDragging,
    buttonProps = _a.buttonProps,
    selectedIcon = _a.selectedIcon,
    onRename = _a.onRename,
    onEditingChange = _a.onEditingChange,
    imageClassName = _a.imageClassName,
    className = _a.className,
    disabled = _a.disabled,
    onFocus = _a.onFocus,
    onBlur = _a.onBlur,
    divProps = tslib_1.__rest(_a, [
      'file',
      'label',
      'selected',
      'dragging',
      'otherDragging',
      'buttonProps',
      'selectedIcon',
      'onRename',
      'onEditingChange',
      'imageClassName',
      'className',
      'disabled',
      'onFocus',
      'onBlur',
    ]);
  var isUserTabbing = (0, hooks_1.useAccessibleFocus)(
    thumbnailFocusObservable_1.thumbnailFocusObservable,
  );
  var _b = (0, hooks_1.useFocus)(onFocus, onBlur),
    focused = _b.focused,
    handleOnFocus = _b.handleOnFocus,
    handleOnBlur = _b.handleOnBlur;
  var _c = (0, hooks_1.useFileSubscribe)(
      file,
      function(f) {
        return f.thumbnail;
      },
      'onthumbnailchange',
    ),
    thumbnail = _c[0],
    thumbnailErr = _c[1];
  var name = (0, hooks_1.useFileSubscribe)(
    file,
    function(f) {
      return f.name;
    },
    'onnamechange',
  )[0];
  var handleOnSave = function(newName) {
    onRename === null || onRename === void 0 ? void 0 : onRename(newName, file);
    onEditingChange === null || onEditingChange === void 0
      ? void 0
      : onEditingChange(false, file);
  };
  var handleOnCancel = function() {
    onEditingChange === null || onEditingChange === void 0
      ? void 0
      : onEditingChange(false, file);
  };
  var handleOnEdit = function() {
    onEditingChange === null || onEditingChange === void 0
      ? void 0
      : onEditingChange(true, file);
  };
  var thumbnailClass = (0, classnames_1.default)(
    'ui__base ui__thumbnail',
    {
      'ui__thumbnail--tabbing': isUserTabbing,
      'ui__thumbnail--selected': selected,
      'ui__thumbnail--focused': focused,
      'ui__thumbnail--disabled': disabled,
      'ui__thumbnail--dragging': dragging,
      'ui__thumbnail--otherDragging': otherDragging,
    },
    className,
  );
  return react_1.default.createElement(
    ClickableDiv_1.ClickableDiv,
    tslib_1.__assign({}, divProps, {
      className: thumbnailClass,
      noFocusStyle: true,
      disabled: disabled,
      onFocus: handleOnFocus,
      onBlur: handleOnBlur,
    }),
    react_1.default.createElement(
      'div',
      { className: 'ui__thumbnail__image' },
      react_1.default.createElement(Image_1.Image, {
        src: thumbnail,
        alt: name,
        pending: !thumbnail && !thumbnailErr,
        onRenderLoading: function() {
          return react_1.default.createElement(FileSkeleton_1.FileSkeleton, {
            className: 'ui__thumbnail__image__skeleton',
          });
        },
        onRenderFallback: function() {
          return react_1.default.createElement(
            FilePlaceholder_1.FilePlaceholder,
            {
              className: 'ui__thumbnail__image__placeholder',
              extension: file.extension,
            },
          );
        },
        className: imageClassName,
      }),
    ),
    react_1.default.createElement(
      'div',
      { className: 'ui__thumbnail__controls' },
      buttonProps === null || buttonProps === void 0
        ? void 0
        : buttonProps.map(function(_a) {
            var key = _a.key,
              buttonPropObject = tslib_1.__rest(_a, ['key']);
            return react_1.default.createElement(
              ToolButton_1.ToolButton,
              {
                key: key,
                disabled: disabled,
                onClick: function(e) {
                  return buttonPropObject.onClick(e, file);
                },
              },
              buttonPropObject.children,
            );
          }),
    ),
    selectedIcon
      ? react_1.default.createElement(
          'div',
          { className: 'ui__thumbnail__selectedIcon' },
          selectedIcon,
        )
      : undefined,
    (label !== null && label !== void 0 ? label : name) || onRename
      ? react_1.default.createElement(EditableText_1.EditableText, {
          className: 'ui__thumbnail__label',
          value: label !== null && label !== void 0 ? label : name,
          centerText: true,
          disabled: disabled,
          locked: !onRename || otherDragging,
          onSave: handleOnSave,
          onCancel: handleOnCancel,
          onEdit: handleOnEdit,
        })
      : undefined,
  );
}
exports.Thumbnail = Thumbnail;
//# sourceMappingURL=Thumbnail.js.map
