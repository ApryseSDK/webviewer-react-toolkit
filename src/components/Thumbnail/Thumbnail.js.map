{"version":3,"file":"Thumbnail.js","sourceRoot":"","sources":["Thumbnail.tsx"],"names":[],"mappings":";;;;AAAA,kEAAoC;AACpC,wDAAgE;AAEhE,qCAA6E;AAC7E,gDAAkE;AAClE,gDAA+C;AAC/C,sDAAqD;AACrD,gDAA+C;AAC/C,kCAAiC;AACjC,4CAA2C;AAC3C,uEAAsE;AA0DtE,SAAgB,SAAS,CAAqB,EAgB1B;IAflB,IAAA,IAAI,UAAA,EACJ,KAAK,WAAA,EACL,QAAQ,cAAA,EACR,QAAQ,cAAA,EACR,aAAa,mBAAA,EACb,WAAW,iBAAA,EACX,YAAY,kBAAA,EACZ,QAAQ,cAAA,EACR,eAAe,qBAAA,EACf,cAAc,oBAAA,EACd,SAAS,eAAA,EACT,QAAQ,cAAA,EACR,OAAO,aAAA,EACP,MAAM,YAAA,EACH,QAAQ,sBAfiC,wLAgB7C,CADY;IAEX,IAAM,aAAa,GAAG,IAAA,0BAAkB,EAAC,mDAAwB,CAAC,CAAC;IAE7D,IAAA,KAA2C,IAAA,gBAAQ,EAAC,OAAO,EAAE,MAAM,CAAC,EAAlE,OAAO,aAAA,EAAE,aAAa,mBAAA,EAAE,YAAY,kBAA8B,CAAC;IAErE,IAAA,KAA4B,IAAA,wBAAgB,EAAC,IAAI,EAAE,UAAC,CAAC,IAAK,OAAA,CAAC,CAAC,SAAS,EAAX,CAAW,EAAE,mBAAmB,CAAC,EAA1F,SAAS,QAAA,EAAE,YAAY,QAAmE,CAAC;IAC3F,IAAA,IAAI,GAAI,IAAA,wBAAgB,EAAC,IAAI,EAAE,UAAC,CAAC,IAAK,OAAA,CAAC,CAAC,IAAI,EAAN,CAAM,EAAE,cAAc,CAAC,GAAzD,CAA0D;IAErE,IAAM,YAAY,GAAG,UAAC,OAAe;QACnC,QAAQ,aAAR,QAAQ,uBAAR,QAAQ,CAAG,OAAO,EAAE,IAAI,CAAC,CAAC;QAC1B,eAAe,aAAf,eAAe,uBAAf,eAAe,CAAG,KAAK,EAAE,IAAI,CAAC,CAAC;IACjC,CAAC,CAAC;IAEF,IAAM,cAAc,GAAG;QACrB,eAAe,aAAf,eAAe,uBAAf,eAAe,CAAG,KAAK,EAAE,IAAI,CAAC,CAAC;IACjC,CAAC,CAAC;IAEF,IAAM,YAAY,GAAG;QACnB,eAAe,aAAf,eAAe,uBAAf,eAAe,CAAG,IAAI,EAAE,IAAI,CAAC,CAAC;IAChC,CAAC,CAAC;IAEF,IAAM,cAAc,GAAG,IAAA,oBAAU,EAC/B,wBAAwB,EACxB;QACE,wBAAwB,EAAE,aAAa;QACvC,yBAAyB,EAAE,QAAQ;QACnC,wBAAwB,EAAE,OAAO;QACjC,yBAAyB,EAAE,QAAQ;QACnC,yBAAyB,EAAE,QAAQ;QACnC,8BAA8B,EAAE,aAAa;KAC9C,EACD,SAAS,CACV,CAAC;IAEF,OAAO,CACL,8BAAC,2BAAY,uBACP,QAAQ,IACZ,SAAS,EAAE,cAAc,EACzB,YAAY,QACZ,QAAQ,EAAE,QAAQ,EAClB,OAAO,EAAE,aAAa,EACtB,MAAM,EAAE,YAAY;QAEpB,uCAAK,SAAS,EAAC,sBAAsB;YACnC,8BAAC,aAAK,IACJ,GAAG,EAAE,SAAS,EACd,GAAG,EAAE,IAAI,EACT,OAAO,EAAE,CAAC,SAAS,IAAI,CAAC,YAAY,EACpC,eAAe,EAAE,cAAM,OAAA,8BAAC,2BAAY,IAAC,SAAS,EAAC,gCAAgC,GAAG,EAA3D,CAA2D,EAClF,gBAAgB,EAAE,cAAM,OAAA,CACtB,8BAAC,iCAAe,IAAC,SAAS,EAAC,mCAAmC,EAAC,SAAS,EAAE,IAAI,CAAC,SAAS,GAAI,CAC7F,EAFuB,CAEvB,EACD,SAAS,EAAE,cAAc,GACzB,CACE;QACN,uCAAK,SAAS,EAAC,yBAAyB,IACrC,WAAW,aAAX,WAAW,uBAAX,WAAW,CAAE,GAAG,CAAC,UAAC,EAA4B;YAA1B,IAAA,GAAG,SAAA,EAAK,gBAAgB,sBAA1B,OAA4B,CAAF;YAAO,OAAA,CAClD,8BAAC,uBAAU,IAAC,GAAG,EAAE,GAAG,EAAE,QAAQ,EAAE,QAAQ,EAAE,OAAO,EAAE,UAAC,CAAC,IAAK,OAAA,gBAAgB,CAAC,OAAO,CAAC,CAAC,EAAE,IAAI,CAAC,EAAjC,CAAiC,IACxF,gBAAgB,CAAC,QAAQ,CACf,CACd,CAAA;SAAA,CAAC,CACE;QACL,YAAY,CAAC,CAAC,CAAC,uCAAK,SAAS,EAAC,6BAA6B,IAAE,YAAY,CAAO,CAAC,CAAC,CAAC,SAAS;QAC5F,CAAC,KAAK,aAAL,KAAK,cAAL,KAAK,GAAI,IAAI,CAAC,IAAI,QAAQ,CAAC,CAAC,CAAC,CAC7B,8BAAC,2BAAY,IACX,SAAS,EAAC,sBAAsB,EAChC,KAAK,EAAE,KAAK,aAAL,KAAK,cAAL,KAAK,GAAI,IAAI,EACpB,UAAU,QACV,QAAQ,EAAE,QAAQ,EAClB,MAAM,EAAE,CAAC,QAAQ,IAAI,aAAa,EAClC,MAAM,EAAE,YAAY,EACpB,QAAQ,EAAE,cAAc,EACxB,MAAM,EAAE,YAAY,GACpB,CACH,CAAC,CAAC,CAAC,SAAS,CACA,CAChB,CAAC;AACJ,CAAC;AA7FD,8BA6FC"}