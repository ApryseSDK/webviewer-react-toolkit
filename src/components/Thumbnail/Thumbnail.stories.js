'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
exports.RotatedThrottled = exports.Rotated = exports.SelectedIcon = exports.WithLabel = exports.WithToolButtons = exports.Rejected = exports.Expensive = exports.Basic = void 0;
var tslib_1 = require('tslib');
var addon_knobs_1 = require('@storybook/addon-knobs');
var react_1 = tslib_1.__importDefault(require('react'));
var action_1 = require('../../storybook-helpers/action/action');
var files_1 = require('../../storybook-helpers/data/files');
var Icon_1 = require('../Icon');
var Thumbnail_1 = require('../Thumbnail');
var README_md_1 = tslib_1.__importDefault(require('./README.md'));
exports.default = {
  title: 'Components/Thumbnail',
  component: Thumbnail_1.Thumbnail,
  parameters: { readme: README_md_1.default },
};
var defaultProps = function(options, index, withToolButtons) {
  if (index === void 0) {
    index = 0;
  }
  return {
    file: (0, files_1.createFile)(index, options),
    selected: (0, addon_knobs_1.boolean)('selected', false),
    disabled: (0, addon_knobs_1.boolean)('disabled', false),
    dragging: (0, addon_knobs_1.boolean)('dragging', false),
    otherDragging: (0, addon_knobs_1.boolean)('otherDragging', false),
    onClick: (0, addon_knobs_1.boolean)('has onClick', true)
      ? (0, action_1.action)('onClick')
      : undefined,
    onRename: (0, addon_knobs_1.boolean)('has onRename', true)
      ? (0, action_1.action)('onRename')
      : undefined,
    buttonProps: withToolButtons
      ? [
          {
            children: react_1.default.createElement(Icon_1.Icon, {
              icon: 'RotateRight',
            }),
            onClick: (0, action_1.action)('rotate onClick'),
            key: 0,
          },
          {
            children: react_1.default.createElement(Icon_1.Icon, {
              icon: 'Close',
            }),
            onClick: (0, action_1.action)('close onClick'),
            key: 1,
          },
        ]
      : undefined,
  };
};
var Basic = function() {
  return react_1.default.createElement(
    Thumbnail_1.Thumbnail,
    tslib_1.__assign({}, defaultProps()),
  );
};
exports.Basic = Basic;
var Expensive = function() {
  return react_1.default.createElement(
    Thumbnail_1.Thumbnail,
    tslib_1.__assign({}, defaultProps({ lazy: true })),
  );
};
exports.Expensive = Expensive;
var Rejected = function() {
  return react_1.default.createElement(
    Thumbnail_1.Thumbnail,
    tslib_1.__assign({}, defaultProps({ error: true })),
  );
};
exports.Rejected = Rejected;
var WithToolButtons = function() {
  return react_1.default.createElement(
    Thumbnail_1.Thumbnail,
    tslib_1.__assign({}, defaultProps(undefined, undefined, true)),
  );
};
exports.WithToolButtons = WithToolButtons;
var WithLabel = function() {
  return react_1.default.createElement(
    Thumbnail_1.Thumbnail,
    tslib_1.__assign(
      { label: (0, addon_knobs_1.text)('label', 'some_label') },
      defaultProps(),
    ),
  );
};
exports.WithLabel = WithLabel;
var SelectedIcon = function() {
  return react_1.default.createElement(
    Thumbnail_1.Thumbnail,
    tslib_1.__assign({}, defaultProps(), {
      selectedIcon: react_1.default.createElement(
        'div',
        { style: { color: 'red', fontSize: 20 } },
        (0, addon_knobs_1.text)('selectedIcon', 'CUSTOM!'),
      ),
    }),
  );
};
exports.SelectedIcon = SelectedIcon;
var Rotated = function() {
  return react_1.default.createElement(
    Thumbnail_1.Thumbnail,
    tslib_1.__assign({}, defaultProps(undefined, 1)),
  );
};
exports.Rotated = Rotated;
var RotatedThrottled = function() {
  return react_1.default.createElement(
    Thumbnail_1.Thumbnail,
    tslib_1.__assign({}, defaultProps({ lazy: true }, 1)),
  );
};
exports.RotatedThrottled = RotatedThrottled;
//# sourceMappingURL=Thumbnail.stories.js.map
