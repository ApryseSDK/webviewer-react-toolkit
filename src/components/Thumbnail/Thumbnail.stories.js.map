{"version":3,"file":"Thumbnail.stories.js","sourceRoot":"","sources":["Thumbnail.stories.tsx"],"names":[],"mappings":";;;;AAAA,sDAAuD;AACvD,wDAA0B;AAC1B,gEAA+D;AAC/D,4DAA6F;AAC7F,gCAA+B;AAC/B,0CAAyD;AACzD,kEAAiC;AAEjC,kBAAe,EAAE,KAAK,EAAE,sBAAsB,EAAE,SAAS,EAAE,qBAAS,EAAE,UAAU,EAAE,EAAE,MAAM,qBAAA,EAAE,EAAE,CAAC;AAE/F,IAAM,YAAY,GAAG,UAAC,OAA2B,EAAE,KAAS,EAAE,eAAyB;IAApC,sBAAA,EAAA,SAAS;IAA0D,OAAA,CAAC;QACrH,IAAI,EAAE,IAAA,kBAAU,EAAC,KAAK,EAAE,OAAO,CAAC;QAChC,QAAQ,EAAE,IAAA,qBAAO,EAAC,UAAU,EAAE,KAAK,CAAC;QACpC,QAAQ,EAAE,IAAA,qBAAO,EAAC,UAAU,EAAE,KAAK,CAAC;QACpC,QAAQ,EAAE,IAAA,qBAAO,EAAC,UAAU,EAAE,KAAK,CAAC;QACpC,aAAa,EAAE,IAAA,qBAAO,EAAC,eAAe,EAAE,KAAK,CAAC;QAC9C,OAAO,EAAE,IAAA,qBAAO,EAAC,aAAa,EAAE,IAAI,CAAC,CAAC,CAAC,CAAC,IAAA,eAAM,EAAC,SAAS,CAAC,CAAC,CAAC,CAAC,SAAS;QACrE,QAAQ,EAAE,IAAA,qBAAO,EAAC,cAAc,EAAE,IAAI,CAAC,CAAC,CAAC,CAAC,IAAA,eAAM,EAAC,UAAU,CAAC,CAAC,CAAC,CAAC,SAAS;QACxE,WAAW,EAAE,eAAe;YAC1B,CAAC,CAAC;gBACE,EAAE,QAAQ,EAAE,8BAAC,WAAI,IAAC,IAAI,EAAC,aAAa,GAAG,EAAE,OAAO,EAAE,IAAA,eAAM,EAAC,gBAAgB,CAAC,EAAE,GAAG,EAAE,CAAC,EAAE;gBACpF,EAAE,QAAQ,EAAE,8BAAC,WAAI,IAAC,IAAI,EAAC,OAAO,GAAG,EAAE,OAAO,EAAE,IAAA,eAAM,EAAC,eAAe,CAAC,EAAE,GAAG,EAAE,CAAC,EAAE;aAC9E;YACH,CAAC,CAAC,SAAS;KACd,CAAC;AAdoH,CAcpH,CAAC;AAEI,IAAM,KAAK,GAAG,cAAM,OAAA,8BAAC,qBAAS,uBAAK,YAAY,EAAE,EAAI,EAAjC,CAAiC,CAAC;AAAhD,QAAA,KAAK,SAA2C;AAEtD,IAAM,SAAS,GAAG,cAAM,OAAA,8BAAC,qBAAS,uBAAK,YAAY,CAAC,EAAE,IAAI,EAAE,IAAI,EAAE,CAAC,EAAI,EAA/C,CAA+C,CAAC;AAAlE,QAAA,SAAS,aAAyD;AAExE,IAAM,QAAQ,GAAG,cAAM,OAAA,8BAAC,qBAAS,uBAAK,YAAY,CAAC,EAAE,KAAK,EAAE,IAAI,EAAE,CAAC,EAAI,EAAhD,CAAgD,CAAC;AAAlE,QAAA,QAAQ,YAA0D;AAExE,IAAM,eAAe,GAAG,cAAM,OAAA,8BAAC,qBAAS,uBAAK,YAAY,CAAC,SAAS,EAAE,SAAS,EAAE,IAAI,CAAC,EAAI,EAA3D,CAA2D,CAAC;AAApF,QAAA,eAAe,mBAAqE;AAE1F,IAAM,SAAS,GAAG,cAAM,OAAA,8BAAC,qBAAS,qBAAC,KAAK,EAAE,IAAA,kBAAI,EAAC,OAAO,EAAE,YAAY,CAAC,IAAM,YAAY,EAAE,EAAI,EAArE,CAAqE,CAAC;AAAxF,QAAA,SAAS,aAA+E;AAE9F,IAAM,YAAY,GAAG,cAAM,OAAA,CAChC,8BAAC,qBAAS,uBACJ,YAAY,EAAE,IAClB,YAAY,EAAE,uCAAK,KAAK,EAAE,EAAE,KAAK,EAAE,KAAK,EAAE,QAAQ,EAAE,EAAE,EAAE,IAAG,IAAA,kBAAI,EAAC,cAAc,EAAE,SAAS,CAAC,CAAO,IACjG,CACH,EALiC,CAKjC,CAAC;AALW,QAAA,YAAY,gBAKvB;AAEK,IAAM,OAAO,GAAG,cAAM,OAAA,8BAAC,qBAAS,uBAAK,YAAY,CAAC,SAAS,EAAE,CAAC,CAAC,EAAI,EAA7C,CAA6C,CAAC;AAA9D,QAAA,OAAO,WAAuD;AAEpE,IAAM,gBAAgB,GAAG,cAAM,OAAA,8BAAC,qBAAS,uBAAK,YAAY,CAAC,EAAE,IAAI,EAAE,IAAI,EAAE,EAAE,CAAC,CAAC,EAAI,EAAlD,CAAkD,CAAC;AAA5E,QAAA,gBAAgB,oBAA4D"}