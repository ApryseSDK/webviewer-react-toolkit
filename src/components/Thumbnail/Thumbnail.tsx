import classnames from 'classnames';
import React, { forwardRef, MouseEvent, useEffect, useImperativeHandle, useRef } from 'react';
import { File } from '../../data/file';
import useFile from '../../hooks/useFile';
import useFocus from '../../hooks/useFocus';
import close from '../../icons/close-24px.svg';
import rotate from '../../icons/rotate_right-24px.svg';
import ClickableDiv, { ClickableDivProps } from '../ClickableDiv';
import EditableText from '../EditableText';
import FileSkeleton from '../FileSkeleton';
import Image from '../Image';
import ToolButton from '../ToolButton';

export interface ThumbnailProps extends ClickableDivProps {
  /**
   * The file to display the thumbnail from.
   */
  file: File;
  /**
   * Optional label. Will use file name if not provided.
   */
  label?: string;
  /**
   * Do not add the file's extension onto the end of the label.
   */
  hideExtension?: boolean;
  /**
   * Display thumbnail with selected props.
   */
  selected?: boolean;
  /**
   * Display this thumbnail as a dragging item.
   */
  dragging?: boolean;
  /**
   * Set to true while other thumbnails are dragging to prevent changes to
   * appearance due to mouse hover.
   */
  otherDragging?: boolean;
  /**
   * Callback fired when the rotate button is clicked. If not given, will have
   * no rotate button.
   */
  onRotate?: (event: MouseEvent<HTMLButtonElement>, file: File) => void;
  /**
   * Callback fired when the remove button is clicked. If not given, will have
   * no remove button
   */
  onRemove?: (event: MouseEvent<HTMLButtonElement>, file: File) => void;
  /**
   * Callback fired when the name is edited and saved.
   */
  onRename?: (newName: string, file: File) => void;
  /**
   * Callback fired whenever edit mode changes.
   */
  onEditingChange?: (isEditing: boolean, file: File) => void;
  /**
   * If provided, can set the delay for fetching any lazy async items from file.
   */
  throttle?: number;
}

export const Thumbnail = forwardRef<HTMLDivElement, ThumbnailProps>(
  (
    {
      file: _file,
      label,
      hideExtension,
      selected,
      dragging,
      otherDragging,
      onRotate,
      onRemove,
      onRename,
      onEditingChange,
      throttle,
      className,
      disabled,
      onFocus,
      onBlur,
      ...divProps
    },
    ref,
  ) => {
    const thumbnailRef = useRef<HTMLDivElement>(null);
    useImperativeHandle(ref, () => thumbnailRef.current!);

    const { focused, handleOnFocus, handleOnBlur } = useFocus(onFocus, onBlur);

    useEffect(() => {
      if (!selected && focused && thumbnailRef.current) {
        thumbnailRef.current.focus();
      }
    }, [focused, selected]);

    const file = useFile(_file, throttle);

    const handleOnSave = (newName: string) => {
      onRename?.(newName, file.file);
      onEditingChange?.(false, file.file);
    };

    const handleOnCancel = () => {
      onEditingChange?.(false, file.file);
    };

    const handleOnEdit = () => {
      onEditingChange?.(true, file.file);
    };

    const onRenderText = (value: string) => {
      if (!value) return '';
      return `${value}.${file.extension}`;
    };

    const thumbnailClass = classnames(
      'ui__base ui__thumbnail',
      {
        'ui__thumbnail--selected': selected,
        'ui__thumbnail--focused': focused,
        'ui__thumbnail--disabled': disabled,
        'ui__thumbnail--dragging': dragging,
        'ui__thumbnail--otherDragging': otherDragging,
      },
      className,
    );

    return (
      <ClickableDiv
        {...divProps}
        className={thumbnailClass}
        ref={thumbnailRef}
        noFocusStyle
        disabled={disabled}
        onFocus={handleOnFocus}
        onBlur={handleOnBlur}
      >
        <div className="ui__thumbnail__image">
          <Image src={file.thumbnail} alt={file.name} onRenderLoading={() => <FileSkeleton />} />
        </div>
        <div className="ui__thumbnail__controls">
          {onRotate ? (
            <ToolButton disabled={disabled} onClick={e => onRotate(e, file.file)} tabIndex={selected ? undefined : -1}>
              <img src={rotate} alt={'rotate'} />
            </ToolButton>
          ) : (
            undefined
          )}
          {onRemove ? (
            <ToolButton disabled={disabled} onClick={e => onRemove(e, file.file)} tabIndex={selected ? undefined : -1}>
              <img src={close} alt={'close'} />
            </ToolButton>
          ) : (
            undefined
          )}
        </div>
        <EditableText
          className="ui__thumbnail__label"
          value={label ?? file.name}
          onRenderText={hideExtension ? undefined : onRenderText}
          centerText
          disabled={disabled}
          locked={!onRename || otherDragging}
          onSave={handleOnSave}
          onCancel={handleOnCancel}
          onEdit={handleOnEdit}
          tabIndex={selected ? undefined : -1}
        />
      </ClickableDiv>
    );
  },
);
