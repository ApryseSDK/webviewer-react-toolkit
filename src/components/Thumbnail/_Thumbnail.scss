$thumbnail-border-size: 2px;

.ui__thumbnail {
  padding: $padding;
  position: relative;
  outline: none;
  width: $thumbnail-dimensions;
  height: $thumbnail-dimensions;
  transition: 0.1s background-color;
  display: flex;
  flex-direction: column;

  &--selected:not(.ui__thumbnail--dragging) {
    background-color: $color-blue-gray-1;
  }

  &--dragging {
    border: $thumbnail-border-size dashed $color-theme-primary;
    z-index: $z-index-dragging;

    margin: -($thumbnail-border-size / 2);

    width: $thumbnail-border-size + $thumbnail-dimensions;
    height: $thumbnail-border-size + $thumbnail-dimensions;

    .ui__thumbnail__controls,
    .ui__thumbnail__image,
    .ui__thumbnail__label {
      visibility: hidden;
    }
  }

  &--otherDragging {
    pointer-events: none;
  }

  @include for-phone-only {
    .ui__thumbnail__controls {
      opacity: 1;
    }
  }

  &:hover,
  &--focused.ui__thumbnail--tabbing:not(.ui__thumbnail--disabled) {
    // Do not style for hover or focus if anything is dragging
    &:not(.ui__thumbnail--dragging):not(.ui__thumbnail--otherDragging) {
      .ui__thumbnail__controls {
        opacity: 1;
      }
    }
  }

  &:hover,
  &--focused.ui__thumbnail--tabbing:not(.ui__thumbnail--disabled) {
    // Do not style for hover or focus if anything is dragging
    &:not(.ui__thumbnail--dragging):not(.ui__thumbnail--otherDragging):not(.ui__thumbnail--selected) {
      background-color: $color-gray-3;
    }
  }

  &:focus.ui__thumbnail--tabbing:not(.ui__thumbnail--disabled) {
    .ui__thumbnail__image {
      img,
      .ui__thumbnail__image__skeleton {
        @include focus;
        border-color: $color-focus-shadow;
      }
    }
  }

  &__controls {
    @include opacity-transition-fast;
    position: absolute;
    top: $padding-tiny;
    right: $padding-tiny;
    opacity: 0;
    display: flex;
    border-radius: $border-radius;
    background-color: $color-background-canvas;
  }

  &__selectedIcon {
    position: absolute;
    top: $padding-tiny;
    left: $padding-tiny;
  }

  &__image {
    @include flex-center;
    flex-direction: column;
    height: 0;
    flex: 1 1 auto;
    pointer-events: none;

    img,
    .ui__thumbnail__image__skeleton {
      border: 1px solid $color-blue-gray-1;
      transition: $focus-transition;
    }
  }

  &__label {
    text-align: center;
    margin-top: $padding;
  }
}
