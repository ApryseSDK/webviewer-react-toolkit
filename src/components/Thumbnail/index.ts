export { Thumbnail as default, ThumbnailProps } from './Thumbnail';
