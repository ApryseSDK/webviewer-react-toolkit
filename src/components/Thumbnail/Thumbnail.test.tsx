import { shallow } from 'enzyme';
import React from 'react';
import { spy } from 'sinon';
import { createFile } from '../../storybook-helpers/data/files';
import { Thumbnail } from '../Thumbnail';

const testFile = createFile(0, { pending: true });

describe('Thumbnail component', () => {
  it('renders its contents', () => {
    const thumbnail = shallow(<Thumbnail file={testFile} />);
    expect(thumbnail.find('.ui__thumbnail')).toHaveLength(1);
  });

  it('snapshot renders default thumbnail', () => {
    const thumbnail = shallow(<Thumbnail file={testFile} />);
    expect(thumbnail).toMatchSnapshot();
  });

  it('clicking thumbnail triggers onClick prop', () => {
    const onClick = spy();
    shallow(<Thumbnail file={testFile} onClick={onClick} />).simulate('click');
    expect(onClick.callCount).toBe(1);
  });
});
