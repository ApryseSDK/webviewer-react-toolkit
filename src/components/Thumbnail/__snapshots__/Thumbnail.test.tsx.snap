// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Thumbnail component snapshot renders default thumbnail 1`] = `
<ForwardRef
  className="ui__base ui__thumbnail"
  noFocusStyle={true}
  onBlur={[Function]}
  onFocus={[Function]}
>
  <div
    className="ui__thumbnail__image"
  >
    <ForwardRef
      alt="file_1.pdf"
      onRenderFallback={[Function]}
      onRenderLoading={[Function]}
      pending={true}
    />
  </div>
  <div
    className="ui__thumbnail__controls"
  />
  <ForwardRef
    centerText={true}
    className="ui__thumbnail__label"
    locked={true}
    onCancel={[Function]}
    onEdit={[Function]}
    onSave={[Function]}
    value="file_1.pdf"
  />
</ForwardRef>
`;
